src/sandbox/hooks/use-auth.tsx

```diff
@@ -0,0 +1,98 @@
+import { useCallback, useEffect, useState } from 'react';
+import useConstant from 'use-constant';
+import debounce from 'awesome-debounce-promise';
+import { useRecoilState, useRecoilValue, useResetRecoilState } from 'recoil';
+import {
+  authOptionsAtom,
+  isSignedInSelector,
+  pendingSignInState,
+  userDataAtom,
+  userSession,
+} from '@sandbox/store/sandbox';
+import { authResponseState } from '@store/auth';
+import type { AuthOptions, FinishedData } from '@stacks/connect';
+import { IS_BROWSER } from '@common/constants';
+import { useRouter } from 'next/router';
+
+export const useAuthState = () => {
+  const isSignedIn = useRecoilValue<boolean>(isSignedInSelector);
+  const [authOptions, setAuthOptions] = useRecoilState<Partial<AuthOptions>>(authOptionsAtom);
+  const [authResponse, setAuthResponse] = useRecoilState(authResponseState);
+  const [userData, setUserData] = useRecoilState(userDataAtom);
+  const resetUserData = useResetRecoilState(userDataAtom);
+
+  const handleSetAuthResponse = (_authResponse: string) => {
+    setAuthResponse(_authResponse);
+  };
+
+  return {
+    isSignedIn: !!userData,
+    userData,
+    setUserData,
+    authOptions,
+    setAuthOptions,
+    userSession: authOptions.userSession,
+    resetUserData,
+    authResponse,
+    setAuthResponse: handleSetAuthResponse,
+  };
+};
+
+export const useAuth = () => {
+  const pendingSignIn = useRecoilValue(pendingSignInState);
+  const { setUserData, authResponse } = useAuthState();
+  const router = useRouter();
+
+  // we are using useConstant and debounce because it seems connect fires this fn many times
```

Comment:
This should be fixed - connect is now much smarter about only calling the callback that initiated a specific request. But maybe it's still an issue, this is fine to keep.